### ISSUE TITLE
proposal: Split PHP side to seperate module

### ISSUE THREAD

**Issue by owenBeresford** [ematsakov/highlight#18]
Hi ,
I have added patches to JS library a few times.
I note that the PHP function names are 50% of the file.  If you don't use PHP at all  this a waste of space.

Propose: split this into "highlight" and "highlight-php" versions in the repo.   
I have no idea how much this library is used as its jQuery.

**Comment by owenBeresford** [comment]
I could offer another patch.

**Comment by ematsakov** [comment]
Hi, @owenBeresford 

Thanks for participating in expanding this plugin.
Sure, feel free to suggest pull request with additional PHP functions list.

> Propose: split this into "highlight" and "highlight-php" versions in the repo.
> I have no idea how much this library is used as its jQuery.

Agree, makes sense to review plugin structure and separate "hightlight_php" logic into sub module. Maybe makes sense to separate all languages into sub modules and lazy load them only if necessary ( in case highlight block exists on the page for corresponding language ).
I'm going to review this library soon and use it on my new site.
Maybe I will get rid of JQuery and rewrite it with pure javascript to keep it as light as possible.

And yes, currently I do not use it a lot :)

Thanks for your interest!

**Comment by owenBeresford** [comment]
I think the choice of keeping jQuery or not is important.  If you look at https://www.npmjs.com/package/jquery-highlight it has 9K downloads/week.   BUT  https://www.npmjs.com/package/highlight.js has >6,000,000 dl/week.   I do not build brochure sites in wordpress, which is where most jQuery is used, I guess.
A github issue isn't the best mechanism for user survey.   
    
It is //just time// to replace the jQuery functions with plain JS, as many features have been added to the language since jQuery started.   Im adverse to committing the time if there is no volume of interest from users.  
 
My boring website is using your library, and will probably keep using it.

**Comment by owenBeresford** [comment]
oh the other suggested thing if doing major works is porting to TS.

**Comment by ematsakov** [comment]
Hi, @owenBeresford 
Not sure about TypeScript, but I'm definitely planning to migrate highlight plugin to native javascript with a lot of improvements in near future.
And will use it on my personal site I just created - https://matsakov.com/ 
Prototype how it will look for PHP syntax highlighting you can  see on homepage of my site. Color scheme is taken from Sublime 3.

Unfortunately, I only can work on this in my spare from work and family time... so not much :)

**Comment by owenBeresford** [comment]
For planning: **https://lor.sh/@tennoseremel/110684294463710459**

**Comment by owenBeresford** [comment]
I just sent you another fix from last year.

Todate I have :
* updated PHP symbols
* updated CSS symbols
* added bash support
* Made numbers in variable names work correctly
* Stopped quotes parsing from mangling everything else
* Stop comments parsing from eating CSS colours
* now stopped function name and strings from colliding with comments parsing
* Fixed a regex

because *I* was the audience who needed it.

Its more sensible to start with a fresh repo if you remove Jquery; but in that case one of the above things in NPM are a pre-packed solution.  A second benefit to the NPM version is the test populations. 

Does Sublime3 have handy syntax files of keywords?  I only mentioned the editor in the previous commits to claim the list I added should be good quality.   It is irrelevant otherwise.

**Comment by ematsakov** [comment]
Thanks for your pull request. Merged.

Yeah, I'll start new repo for 2nd version and get rid from JQuery for sure.

> Its more sensible to start with a fresh repo if you remove Jquery; but in that case one of the above things in NPM is a pre-packed solution. A second benefit to the NPM version is the test populations.

Releasing on npmjs.com makes sense. But in this case it will better to implement package which will work both on client and server ( node ) side to be able to use in frameworks like NextJS which supports server side rendering. Not sure I have a lot of time for this now. And naming is open question. As package with the same name and functionality exists https://www.npmjs.com/package/highlight.js

> Does Sublime3 have handy syntax files of keywords? I only mentioned the editor in the previous commits to claim the list I added should be good quality. Its irrelevant otherwise.

As I know, Sublime packages are in binary format. So not sure if they have all possible keywords or parse code according to some rules.

**Comment by owenBeresford** [comment]
This is still on my #TODOLIST however, if/when I do this:
* I will swap to CSS counter based line numbers, and add support for not starting the number at 0.
* Look at writing the content in a SVG doc, rather than traditional HTML; it should handle having 1000+ new nodes better. Secondly taking the text into a different document means it will not cause any issues with SEO (Google PageRank routinely complains my pages are too large). 
* The fact the quote/ sample is less visible to basic tools shouldn't be a problem as there is still the plaintext version inside the extra render; and the Noscript is still available for the original text.     
* +TODO +UPDATETHISLINE #testing on this plan, "how to vet SVG docs for expected values"
* +TODO +UPDATETHISLINE screenreaders and code highlighters , which is not covered in old version
* +TODO +UPDATETHISLINE screen layout for mobile and code samples. Not covered in old version.  Maybe a new named tab with a very clear back button.  Mobile + code samples is not a clear user journey

**Comment by owenBeresford** [comment]
Some tech I have been waiting on:
* https://www.bram.us/2024/02/18/custom-highlight-api-for-syntax-highlighting/
* https://developer.mozilla.org/en-US/docs/Web/CSS/::highlight

... simple DOM

**Comment by owenBeresford** [comment]
I built a different solution, using highlight.js.   I have dropped jQuery (250KB JS down to 9KB JS) ad am using new test frameworks and TS.

https://github.com/owenBeresford/extra-panda/
